### ISSUE TITLE
weights of embed_tokens and lm_head stay unchanged after training.

### ISSUE THREAD

**Issue by staticpunch** [ymcui/Chinese-LLaMA-Alpaca#604]
### Check before submitting issues

- [X] Make sure to pull the latest code, as some issues and bugs have been fixed.
- [X] Due to frequent dependency updates, please ensure you have followed the steps in our [Wiki](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki)
- [X] I have read the [FAQ section](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki/FAQ) AND searched for similar issues and did not find a similar problem or solution
- [X] Third-party plugin issues - e.g., [llama.cpp](https://github.com/ggerganov/llama.cpp), [text-generation-webui](https://github.com/oobabooga/text-generation-webui), [LlamaChat](https://github.com/alexrozanski/LlamaChat), we recommend checking the corresponding project for solutions
- [X] Model validity check - Be sure to check the model's [SHA256.md](https://github.com/ymcui/Chinese-LLaMA-Alpaca/blob/main/SHA256.md). If the model is incorrect, we cannot guarantee its performance

### Type of Issue

Model inference

### Base Model

None

### Operating System

Linux

### Describe your issue in detail

I try to run the code `run_clm_pt_with_peft.py` on the model `openlm-research/open_llama_3b` with its original tokenizer (not extended). 
- First, since I have only a single GPU and I got error running the script with `deepspeed`, I modified the script  to exclude `deepspeed` parts. Details below: 
```
lr=2e-4
lora_rank=8
lora_alpha=32
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

pretrained_model=openlm-research/open_llama_3b
chinese_tokenizer_path=openlm-research/open_llama_3b
dataset_dir=sample_data
data_cache=temp_data_cache_dir
per_device_train_batch_size=1
per_device_eval_batch_size=1
training_steps=100
gradient_accumulation_steps=1
output_dir=output_dir

python run_clm_pt_with_peft.py \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --data_cache_dir ${data_cache} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --seed $RANDOM \
    --fp16 \
    --max_steps ${training_steps} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.05 \
    --weight_decay 0.01 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --save_steps 500 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --block_size 512 \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --modules_to_save ${modules_to_save} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --gradient_checkpointing \
    --ddp_find_unused_parameters False
```
- Second, I faced the issue `ValueError : Attemting to unscale fp16 Gradients`. After debugging, I found that the reason for the error is that the weights of modules in `modules_to_save`, namely `lm_head` and `embed_tokens`, are in float16. So I tried to convert them to float32 with the following snippet and it ran fine after that. 
```
def convert_trainable_parameters_to_float32(model, modules_to_save):
    """
    Make sure all trainable paramaters are in float32. 
    Otherwise, the training would crash because LoraConfig does not 
    automatically convert float16 -> float32.
    """
    if not modules_to_save: # convert None to [] to make it iterable
        modules_to_save = []
    for name, param in model.named_parameters():
        # last_name = name.split(".")[-1]
        if (any(x in name for x in modules_to_save) and param.requires_grad):
            param.data = param.data.float()
```

- For now, I think the script is fine since the number of trainable params is reasonable (about 11%) the loss decreases stably. However after finishing training for some steps, I tried to load checkpoints and face the issue mentioned in the title:  weights of embed_tokens and lm_head stay unchanged after training. Here is my script to check whether the trained weights changed or not. 
```
base_model = "openlm-research/open_llama_3b"
torch_dtype = torch.float16 
device_map = {"": int(os.environ.get("CUDA_DEVICE") or 0)}
config = AutoConfig.from_pretrained(base_model)
model = LlamaForCausalLM.from_pretrained(
    base_model,
    config=config,
    torch_dtype=torch_dtype,
    device_map = device_map,
    low_cpu_mem_usage=True,
)
old_lm_head = model.lm_head.weight.data.clone()

peft_path = "../Chinese-LLaMA-Alpaca/scripts/training/output_dir/pt_lora_model/"
peft_model = PeftModel.from_pretrained(model, peft_path)
new_lm_head = peft_model.base_model.model.lm_head.original_module.weight.data

print(old_lm_head == new_lm_head)
```
The output of this script is a matrix with all `True` elements. This does not make sense to me because the trained weights should be different from the original ones. 

### Dependencies (must be provided for code-related issues)

absl-py==1.4.0
accelerate @ git+https://github.com/huggingface/accelerate.git@50eabe5b1d3cde24a9f3b65b6e7e25075b269da4
aiohttp==3.8.4
aiosignal==1.3.1
anyio==3.7.0
argon2-cffi==21.3.0
argon2-cffi-bindings==21.2.0
arrow==1.2.3
asttokens==2.2.1
async-lru==2.0.2
async-timeout==4.0.2
attrs==23.1.0
Babel==2.12.1
backcall==0.2.0
beautifulsoup4==4.12.2
bitsandbytes==0.39.0
bleach==6.0.0
cachetools==5.3.1
cchardet==2.1.7
certifi==2023.5.7
cffi==1.15.1
chardet==5.1.0
charset-normalizer==3.1.0
cmake==3.26.4
comm==0.1.3
datasets==2.12.0
debugpy==1.6.7
decorator==5.1.1
defusedxml==0.7.1
dill==0.3.6
exceptiongroup==1.1.1
executing==1.2.0
fastjsonschema==2.17.1
filelock==3.12.2
fire==0.5.0
fqdn==1.5.1
frozenlist==1.3.3
fsspec==2023.6.0
google-auth==2.20.0
google-auth-oauthlib==1.0.0
grpcio==1.54.2
huggingface-hub==0.15.1
idna==3.4
iniconfig==2.0.0
ipykernel==6.23.2
ipython==8.14.0
ipython-genutils==0.2.0
ipywidgets==8.0.6
isoduration==20.11.0
jedi==0.18.2
Jinja2==3.1.2
joblib==1.2.0
json5==0.9.14
jsonpointer==2.3
jsonschema==4.17.3
jupyter==1.0.0
jupyter-console==6.6.3
jupyter-events==0.6.3
jupyter-lsp==2.2.0
jupyter_client==8.2.0
jupyter_core==5.3.0
jupyter_server==2.6.0
jupyter_server_terminals==0.4.4
jupyterlab==4.0.2
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.7
jupyterlab_server==2.23.0
lit==16.0.5.post0
Markdown==3.4.3
MarkupSafe==2.1.3
matplotlib-inline==0.1.6
mistune==2.0.5
mpmath==1.3.0
multidict==6.0.4
multiprocess==0.70.14
nbclassic==1.0.0
nbclient==0.8.0
nbconvert==7.5.0
nbformat==5.9.0
nest-asyncio==1.5.6
networkx==3.1
ninja==1.11.1
notebook==6.5.4
notebook_shim==0.2.3
numpy==1.24.3
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
oauthlib==3.2.2
overrides==7.3.1
packaging==23.1
pandas==2.0.2
pandocfilters==1.5.0
parso==0.8.3
peft @ git+https://github.com/huggingface/peft.git@eb75374fb12aedb369c6bbc2c153047720d4b38d
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.3.0
platformdirs==3.5.3
pluggy==1.0.0
prometheus-client==0.17.0
prompt-toolkit==3.0.38
protobuf==3.20.3
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
pyarrow==12.0.1
pyasn1==0.5.0
pyasn1-modules==0.3.0
pycparser==2.21
Pygments==2.15.1
pyrsistent==0.19.3
pytest==7.3.2
python-dateutil==2.8.2
python-json-logger==2.0.7
pytorch-triton==2.1.0+440fd1bf20
pytz==2023.3
PyYAML==6.0
pyzmq==25.1.0
qtconsole==5.4.3
QtPy==2.3.1
regex==2023.6.3
requests==2.31.0
requests-oauthlib==1.3.1
responses==0.18.0
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rsa==4.9
safetensors==0.3.1
scikit-learn==1.2.2
scipy==1.10.1
Send2Trash==1.8.2
sentencepiece==0.1.99
six==1.16.0
sniffio==1.3.0
soupsieve==2.4.1
stack-data==0.6.2
sympy==1.12
tensorboard==2.13.0
tensorboard-data-server==0.7.1
termcolor==2.3.0
terminado==0.17.1
threadpoolctl==3.1.0
tinycss2==1.2.1
tokenizers==0.13.3
tomli==2.0.1
torch==2.1.0.dev20230613+cu121
torchaudio==2.1.0.dev20230613+cu121
torchvision==0.16.0.dev20230613+cu121
tornado==6.3.2
tqdm==4.65.0
traitlets==5.9.0
transformers @ git+https://github.com/huggingface/transformers.git@91b62f5a785264b626ec7ba708c894ac0276f5cb
triton==2.0.0
typing_extensions==4.6.3
tzdata==2023.3
uri-template==1.2.0
urllib3==1.26.16
wcwidth==0.2.6
webcolors==1.13
webencodings==0.5.1
websocket-client==1.5.3
Werkzeug==2.3.6
widgetsnbextension==4.0.7
xxhash==3.2.0
yarl==1.9.2

**Comment by iMountTai** [comment]
1. Our code recommends install [Peft commit id 13e53fc](https://github.com/huggingface/peft/tree/13e53fc).
2. For the new version of PEFT, duplicate the weights of "embed" and "lm_head." It might be like this： `new_lm_head = peft_model.base_model.model.lm_head.modules_to_save.******`
